source/arm9/math.c

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: Zlib
+//
+// Copyright (C) 2024 Dominik Kurz
+#include <math.h>
+
+
+float hw_sqrtf(float x)
+{
+    union{float f; u32 i;}xu;
+    xu.f = x;
+    u32 mantissa32 = xu.i & ((1 << 23) - 1); // extracts mantissa via bitmask
+    mantissa32 += (1<<23); // adds implicit bit to mantissa.
+    // if you were planning to handle subnormal numbers you would not add this.
+    u64 mantissa= mantissa32; //cast so that the shift doesnt go wrong 
+    mantissa <<= (((xu.i & (1<<23)) == 0) + 25);
+    // applies additional shift depending on whether exponent is even or odd
+    // 25 is 23+2, the +2 is necessary for rounding
+    REG_SQRTCNT = SQRT_64;  
+    REG_SQRT_PARAM = mantissa;
+    // starts hardware squareroot 
+    u32 raw_exponent = (xu.i & (0xff << 23));
+    // check if exponent is 0
+    if (raw_exponent > 0) 
+    {
+        s32 exponent = raw_exponent-(127<<23);  // subtract bias from exponent
+        exponent = exponent >> 1; 
+        // right shift on signed number depends on compiler
+        // but it works in gcc
```

Comment:
I am skeptical of this, especially as I test with Clang from time to time. If it's a GCC-ism, it should be put behind a GCC-specific `#ifdef`.

Reply:
I found a comment on stack exchange stating that arithmetic right shift and integer division have different rounding behaviours, so maybe division wouldn't work.  Would probably have to test. Wikipedia says almost every compiler does arithmetic shifts though, so it'll probably work in clang.

Reply:
![image](https://github.com/blocksds/libnds/assets/47481645/7b94471a-9cd8-46e8-ad6f-a71212bd8c76)
seems /2 gets converted to an ASR instruction on nds

Reply:
To be fair, we already rely on the fact that ">>" on signed integers is an arithmetic shift in other places, like the fixed point math functions. I'm pretty sure this is the kind of things that Clang would never implement in a different way than GCC.

Reply:
I'll change it back to a shift then.  There are cases where gcc will emit division instructions with division by 2 for some optimization levels and for some numbers. I'll just explain it in a comment.